anymail/webhooks/mailpace.py

```diff
@@ -0,0 +1,145 @@
+            )
+            self.warn_if_no_basic_auth = False
+        except AnymailConfigurationError:
+            self.webhook_key = None
+            self.warn_if_no_basic_auth = True
+
+        super().__init__(**kwargs)
+
+    # Used by base class
+    signal = tracking
+
+    event_record_types = {
+        # Map MailPace event RecordType --> Anymail normalized event type
+        "email.queued": EventType.QUEUED,
+        "email.delivered": EventType.DELIVERED,
+        "email.deferred": EventType.DEFERRED,
+        "email.bounced": EventType.BOUNCED,
+        "email.spam": EventType.REJECTED,
+    }
+
+    # MailPace doesn't send a signature for inbound webhooks, yet
+    # When/if MailPace does this, move this to the parent class
+    def validate_request(self, request):
+        if self.webhook_key:
+            try:
+                signature_base64 = request.headers["X-MailPace-Signature"]
+                signature = base64.b64decode(signature_base64)
+            except (KeyError, binascii.Error):
+                raise AnymailWebhookValidationFailure(
+                    "MailPace webhook called with invalid or missing signature"
+                )
+
+            verify_key_base64 = self.webhook_key
+
+            verify_key = VerifyKey(base64.b64decode(verify_key_base64))
+
+            message = request.body
+
+            try:
+                verify_key.verify(message, signature)
+            except (CryptoError, ValueError):
+                raise AnymailWebhookValidationFailure(
+                    "MailPace webhook called with incorrect signature"
+                )
+        else:
+            return True
+            # No webhook key set
+
+    def esp_to_anymail_event(self, esp_event):
+        event_type = self.event_record_types.get(esp_event["event"], EventType.UNKNOWN)
+        payload = esp_event["payload"]
+
+        reject_reason = (
+            RejectReason.SPAM
+            if event_type == EventType.REJECTED
+            else RejectReason.BOUNCED
+            if event_type == EventType.BOUNCED
+            else None
+        )
+        tags = payload.get("tags", [])
+
+        return AnymailTrackingEvent(
+            event_type=event_type,
+            timestamp=parse_datetime(payload["created_at"]),
+            event_id=payload["id"],
+            message_id=payload["message_id"],
+            recipient=payload["to"],
+            tags=tags,
+            reject_reason=reject_reason,
+        )
+
+
+class MailPaceInboundWebhookView(MailPaceBaseWebhookView):
+    """Handler for MailPace inbound webhook"""
+
+    signal = inbound
+
+    def esp_to_anymail_event(self, esp_event):
+        # Use Raw MIME based on guidance here:
+        # https://github.com/anymail/django-anymail/blob/main/ADDING_ESPS.md
+        message = AnymailInboundMessage.parse_raw_mime(esp_event.get("raw", None))
+
+        return AnymailInboundEvent(
+            event_type=EventType.INBOUND,
+            timestamp=timezone.now(),
```

Comment:
Please use `timestamp=None` if the inbound payload doesn't provide an event timestamp. (Timestamp needs to be unchanged if the same event is received twice.)